File: src/nl/rubensten/texifyidea/inspections/latex/LatexSpellcheckingStrategy.java
package nl.rubensten.texifyidea.inspections.latex;

import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.spellchecker.tokenizer.SpellcheckingStrategy;
import com.intellij.spellchecker.tokenizer.Tokenizer;
import nl.rubensten.texifyidea.LatexLanguage;
import nl.rubensten.texifyidea.lang.Argument;
import nl.rubensten.texifyidea.lang.Argument.Type;
import nl.rubensten.texifyidea.lang.LatexMathCommand;
import nl.rubensten.texifyidea.lang.LatexNoMathCommand;
import nl.rubensten.texifyidea.psi.LatexBeginCommand;
import nl.rubensten.texifyidea.psi.LatexCommands;
import nl.rubensten.texifyidea.psi.LatexEndCommand;
import nl.rubensten.texifyidea.psi.LatexTypes;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Optional;

/**
 * @author Ruben Schellekens
 */
public class LatexSpellcheckingStrategy extends SpellcheckingStrategy {

    @Override
    public boolean isMyContext(@NotNull PsiElement psiElement) {
        return psiElement.getLanguage().equals(LatexLanguage.INSTANCE);
    }

    @NotNull
    @Override
    public Tokenizer getTokenizer(PsiElement psiElement) {
        if (psiElement instanceof LeafPsiElement) {
            LeafPsiElement leaf = (LeafPsiElement)psiElement;

            if (leaf.getElementType().equals(LatexTypes.COMMAND_TOKEN) ||
                    leaf.getElementType().equals(LatexTypes.COMMENT_TOKEN) ||
                    isBeginEnd(psiElement)) {
                return EMPTY_TOKENIZER;
            }

            Argument argument = getArgument(leaf);

            if (argument == null && leaf.getElementType().equals(LatexTypes.NORMAL_TEXT)) {
                return TEXT_TOKENIZER;
            }

            if (argument == null || argument.getType() == Type.TEXT) {
                return TEXT_TOKENIZER;
            }
        }

        return EMPTY_TOKENIZER;
    }

    private boolean isBeginEnd(PsiElement element) {
        PsiElement elt = PsiTreeUtil.getParentOfType(element, LatexBeginCommand.class);
        if (elt != null) {
            return true;
        }

        elt = PsiTreeUtil.getParentOfType(element, LatexEndCommand.class);
        return elt != null;
    }

    private Argument getArgument(LeafPsiElement leaf) {
        LatexCommands parent = PsiTreeUtil.getParentOfType(leaf, LatexCommands.class);
        if (parent == null) {
            return null;
        }

        Argument[] arguments = getArguments(parent.getCommandToken().getText().substring(1));
        if (arguments == null) {
            return null;
        }

        List<String> realParams = parent.getRequiredParameters();
        int parameterIndex = realParams.indexOf(leaf.getText());
        if (parameterIndex < 0 || parameterIndex >= arguments.length) {
            return null;
        }

        return arguments[parameterIndex];
    }

    private Argument[] getArguments(String commandName) {
        Optional<LatexNoMathCommand> cmdHuh = LatexNoMathCommand.get(commandName);
        if (cmdHuh.isPresent()) {
            return cmdHuh.get().getArguments();
        }

        LatexMathCommand mathCmdHuh = LatexMathCommand.get(commandName);
        if (mathCmdHuh == null) {
            return null;
        }

        return mathCmdHuh.getArguments();
    }
}

File: src/nl/hannahsten/texifyidea/inspections/LatexSpellcheckingStrategy.kt
package nl.hannahsten.texifyidea.inspections

import com.intellij.lang.LanguageParserDefinitions
import com.intellij.psi.PsiElement
import com.intellij.psi.impl.source.tree.LeafPsiElement
import com.intellij.psi.util.parentOfType
import com.intellij.spellchecker.inspections.SpellCheckingInspection
import com.intellij.spellchecker.tokenizer.SpellcheckingStrategy
import com.intellij.spellchecker.tokenizer.Tokenizer
import nl.hannahsten.texifyidea.grammar.LatexLanguage
import nl.hannahsten.texifyidea.lang.LArgument
import nl.hannahsten.texifyidea.lang.LatexContexts
import nl.hannahsten.texifyidea.lang.predefined.AllPredefined
import nl.hannahsten.texifyidea.psi.*
import nl.hannahsten.texifyidea.settings.TexifySettings
import nl.hannahsten.texifyidea.util.parser.firstParentOfType
import nl.hannahsten.texifyidea.util.parser.hasParent
import nl.hannahsten.texifyidea.util.parser.lookupCommandPsi

/**
 * @author Hannah Schellekens
 */
class LatexSpellcheckingStrategy : SpellcheckingStrategy() {

    override fun isMyContext(psiElement: PsiElement): Boolean = psiElement.language == LatexLanguage

    override fun getTokenizer(psiElement: PsiElement): Tokenizer<*> {
        if (psiElement !is LeafPsiElement) {
            return EMPTY_TOKENIZER
        }

        if (psiElement.elementType == LatexTypes.COMMAND_TOKEN ||
            psiElement.elementType == LatexTypes.COMMAND_IFNEXTCHAR ||
            psiElement.elementType == LatexTypes.LEFT ||
            psiElement.elementType == LatexTypes.RIGHT ||
            isBeginEnd(psiElement)
        ) {
            return EMPTY_TOKENIZER
        }

        // Allow comments to be spellchecked, depending on inspection settings
        if (psiElement.elementType == LatexTypes.COMMENT_TOKEN) {
            return TEXT_TOKENIZER
        }

        return TEXT_TOKENIZER
    }

    /**
     * We need more fine-grained control than just the element types provided in LatexParserDefinition.
     *
     * Literals: normal text and arguments with a known text type, should be on by default
     * Code: anything else that could make sense to spellcheck, should be off by default
     */
    override fun elementFitsScope(psiElement: PsiElement, scope: Set<SpellCheckingInspection.SpellCheckingScope?>?): Boolean {
        if (scope == null || psiElement !is LeafPsiElement) {
            return super.elementFitsScope(psiElement, scope)
        }

        val elementType = psiElement.node.elementType
        val latexParserDefinition = LanguageParserDefinitions.INSTANCE.forLanguage(LatexLanguage)

        // Literals
        val hasLiteralsScope = scope.contains(SpellCheckingInspection.SpellCheckingScope.Literals)
        val argument = getArgument(psiElement)
        // Normal text
        if (
            argument == null &&
            psiElement.elementType == LatexTypes.NORMAL_TEXT_WORD &&
            // Exclude text in parameters by default, unless we know it contains text (e.g. \section)
            !psiElement.hasParent(LatexParameterText::class)
        ) {
            return hasLiteralsScope
        }

        if (argument?.contextSignature?.introduces(LatexContexts.Text) == true) {
            return hasLiteralsScope
        }

        // Comments
        if (latexParserDefinition.commentTokens.contains(elementType)) {
            return scope.contains(SpellCheckingInspection.SpellCheckingScope.Comments)
        }

        // Code
        if (scope.contains(SpellCheckingInspection.SpellCheckingScope.Code)) {
            // Override the default to disable spellchecking for code unless enabled in TeXiFy settings
            return TexifySettings.getState().enableSpellcheckEverywhere
        }

        return false
    }

    private fun isBeginEnd(element: PsiElement): Boolean = element.parentOfType<LatexEndCommand>() != null || element.parentOfType<LatexBeginCommand>() != null

    /**
     * Get the argument (with type) from TeXiFy knowledge that corresponds with the current psi element.
     */
    private fun getArgument(leaf: LeafPsiElement): LArgument? {
        val parameter = leaf.parentOfType<LatexParameter>() ?: return null
        val parent = parameter.parentOfType<LatexCommands>() ?: return null
        val arguments = AllPredefined.lookupCommandPsi(parent)?.arguments ?: return null

        parameter.requiredParam?.let { requiredParam ->
            val idx = parent.indexOfRequiredParameter { it === requiredParam }
            LArgument.getRequiredByIdx(arguments, idx)?.let {
                return it
            }
        }

        // Note that a leaf may be only part of a parameter
        val parameterText = leaf.firstParentOfType(LatexParameterText::class)?.text
        val optionalArguments = arguments.filter { it.isOptional }
        // Also check optional arguments, if not key-value pairs it may contain text
        val optionalParamIndex = parent.getOptionalParameterMap().map { it.key.text }.indexOf(parameterText)
        return if (optionalParamIndex >= 0 && optionalParamIndex < optionalArguments.size) {
            optionalArguments[optionalParamIndex]
        }
        else {
            null
        }
    }
}
